### DISCUSSION TITLE
Can't view the corresponding text from font file

### DISCUSSION THREAD

**Discussion by CodePythonFollow** [pymupdf/PyMuPDF#2517] [Q&A] [answered]
I extracted the fonts using the method below
```python
fonts = page.get_fonts(full=True)

for font in fonts:
    res = file.extract_font(font[0])
```
I use the following method to obtain text information
```python
blocks = page.get_text("dict", flags=11)["blocks"]
for b in blocks:  # iterate through the text blocks
    for l in b["lines"]:  # iterate through the text lines
        for s in l["spans"]:  # iterate through the text spans
            print("")
            font_properties = "Font: '%s' (%s), size %g, color #%06x" % (
                s["font"],  # font name
                flags_decomposer(s["flags"]),  # readable font flags
                s["size"],  # font size
                s["color"],  # font color
            )
            print("Text: '%s'" %  s["text"])  # simple print of text
            print(f"box: {s['bbox']}")  # simple print of text
            print(font_properties)
```
out:
![image](https://github.com/pymupdf/PyMuPDF/assets/52147782/f0946843-5574-44a4-90cd-2acbbc719ccf)

I can't see the corresponding text in the extracted font.  font type: cid 
![image](https://github.com/pymupdf/PyMuPDF/assets/52147782/b662b1dc-0bf6-4d71-83d9-3518157dd032)

The text "重要提示" not in font file.

I wonder if something went wrong with me？

**Comment by JorjMcKie** [comment]
This is not bug, but a "Discussions" item. Let me transfer first.

**Reply by ousia** [reply]
Do you really think this is an announcement? I think Q&A would make more sense here (writing that after mischaracterizing myself my previous discussion as announcement) .

**Reply by JorjMcKie** [reply]
of course you are right!

**Comment by JorjMcKie** [comment]
**_Back-translation_** information from a visible "glyph" on a page to its unicode, that was used to generate that glyph, may or may not be provided by a font - this is voluntary.
1. This information **may be missing** completely: then no text extraction is possible at all and you will only see invalid unicode symbols U+FFFD, '�' (white question mark with a black diamond shaped background).
2. This information **may be incomplete**: only some characters have been forgotten and therefore appear as �.
3. This information **may be given, but wrong on purpose**: this can is used to prevent text extraction. By not letting � appear, a program may not realize an extraction problem (although it is extractiing garbage) and will not invoke OCR - as it would otherwise.

**Comment by CodePythonFollow** [comment]
I have encountered the ' �' character you said, the above situation is a normal display Chinese, but the font file can only see the numbers without getting the Chinese characters.
Are you talking about incomplete font files？

� like it:
![1688570910233](https://github.com/pymupdf/PyMuPDF/assets/52147782/97e6276d-2c7a-4fd0-8d7b-85d3a2bdde93)

**Reply by JorjMcKie** [reply, marked as answer]
> Are you talking about incomplete font files？

Yes, exactly. Could be that the font in question only has `/ToUnicode` entries for numbers and maybe always returns the space character code for all other glyphs.

WRT to "incomplete" fonts: The `/ToUnicode` information is optional for a font. There is no law or rule prescribing that a font must provide this information at all. Please recall that PDF has originally been created to display information for human reception. Not as a data store - things like text extraction, image extraction, etc. came later and - as I wrote - are not necessarily reliable.

**Comment by CodePythonFollow** [comment]
Thank you very much.
